src/main/java/org/truffleruby/core/thread/ThreadManager.java

```diff
@@ -195,11 +195,15 @@ public static void run(DynamicObject thread, RubyContext context, Node currentNo
             task.run();
         } catch (ThreadExitException e) {
             setThreadValue(context, thread, context.getCoreLibrary().getNil());
-            return;
         } catch (RaiseException e) {
             setException(context, thread, e.getException(), currentNode);
         } catch (ReturnException e) {
             setException(context, thread, context.getCoreExceptions().unexpectedReturn(currentNode), currentNode);
+        } catch (ExitException e) {
+            final Thread rootThread = Layouts.FIBER.getThread(Layouts.THREAD.getFiberManager(context.getThreadManager().getRootThread()).getCurrentFiber());
+            context.getSafepointManager().pauseThreadAndExecute(rootThread, currentNode, (actionThread, actionCurrentNode) -> {
+                throw e;
+            });
```

Comment:
Because I do not want complex logic in the middle of this already long catch handler,
and it's not clear from the code what it does so a method would document it by the name.